pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,344 @@
+	}
+
+	bulkResponse, err := bulkRequest.Do(ctx)
+	if err != nil {
+		return err
+	}
+
+	if bulkResponse != nil {
+
+	}
+	return nil
+}
+
+// readBatch represents a batch of rows read from ElasticSearch.
+type readBatch struct {
+	rangeValue []byte
+	value      []byte
+}
+
+func (r readBatch) Iterator() chunk.ReadBatchIterator {
+	return &elasticReadBatchIterator{
+		readBatch: r,
+	}
+}
+
+type elasticReadBatchIterator struct {
+	consumed bool
+	readBatch
+}
+
+func (r *elasticReadBatchIterator) Next() bool {
+	if r.consumed {
+		return false
+	}
+	r.consumed = true
+	return true
+}
+
+func (r *elasticReadBatchIterator) RangeValue() []byte {
+	return r.rangeValue
+}
+
+func (r *elasticReadBatchIterator) Value() []byte {
+	return r.value
+}
+
+func (e *esClient) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error {
+	return chunk_util.DoParallelQueries(ctx, e.query, queries, callback)
+}
+
+func (e *esClient) query(ctx context.Context, query chunk.IndexQuery, callback func(chunk.ReadBatch) (shouldContinue bool)) error {
+	var rangeQuery *elastic.RangeQuery
+	var valueTermQuery *elastic.TermQuery
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf(
+		"hash [%s], rangeValuePrefix [%s], rangeValueStart [%s]", query.HashValue, query.RangeValuePrefix, query.RangeValueStart))
+	hashTermQuery := elastic.NewTermQuery("hash", query.HashValue)
+	switch {
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual != nil:
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual == nil:
+		break
+	}
+
+	exists, err := e.client.IndexExists(query.TableName).Do(ctx)
+	if err != nil {
+		// Handle error
+		fmt.Println(exists)
+	}
+
+	// Search with a term query
+	baseQuery := e.client.Search().
+		Index(query.TableName).
+		Query(hashTermQuery)
+	if valueTermQuery != nil {
+		baseQuery = baseQuery.Query(valueTermQuery)
+	}
+	if rangeQuery != nil {
+		baseQuery = baseQuery.Query(rangeQuery)
+	}
+
+	// Search with a term query
+	searchResult, err := baseQuery.
+		Sort("range", true). // sort by "range" field, ascending
+		From(0).Size(maxFetchDocs).
```

Comment:
There seems to be a fixed limit here of 1000 records.

Reply:
Will change it to configurable.

Reply:
While this is now configurable, your design is limited to return only that number of results total. Also, it cannot be larger than `index.max_result_window` which by default is set to 10K in ES. 

In order to be an effective at scale index store, the storage layer needs to be able to return many tens of thousands of results. You probably need to look at this search portion of code and add pagination for larger result sets, otherwise, there is a decent risk of results being truncated from anything but small result sets.